app/models/spree/payment/processing.rb

```diff
@@ -232,6 +232,8 @@ def handle_response(response, success_state, failure_state)
           unless response.authorization.nil?
             self.response_code = response.authorization
             self.avs_response = response.avs_result['code']
+            self.captured_at = Time.zone.now if success_state == :complete &&
```

Comment:
The main question I have here is: is this the only place in the entire codebase where a payment can be captured? I'm not sure! :sweat_smile:

Reply:
Hmmm.... also I think the use of `success_state` here isn't quite right. The method takes two arguments `success_state` and `failed_state`, but they're used to indicate what state the payment should be changed to in either case. I think the conditional above (`if response.success?`) is the bit that actually indicates what has happened.

Hmmm... but actually I think this isn't the right place to put this bit of logic. I would hook it into the `state_machines` definition in the `Spree::Payment` model (around [line 92](https://github.com/openfoodfoundation/openfoodnetwork/blob/0cc7bdeafa38802d1a0486456faa57c41c91e97a/app/models/spree/payment.rb#L92)) with:

```ruby
after_transition to: :complete, do: :set_captured_at
```

and then define that `set_captured_at` as a private method in `Spree::Payment`. There's some examples of this in the state machines definition in the [Checkout module](https://github.com/openfoodfoundation/openfoodnetwork/blob/0cc7bdeafa38802d1a0486456faa57c41c91e97a/app/models/spree/order/checkout.rb#L70-L76).

Ultimately when any transitions happen they trigger a state change, and that's where we want to add this :+1:

Reply:
@Matt-Yorkley Thanks, using `after_transition` sounds better, it's fixed up now. I had to do the `success_state == :complete` before to stop it from being set for `authorize` calls too.